plugins/integrations/prometheus/src/main/java/org/apache/cloudstack/metrics/PrometheusExporterImpl.java

```diff
@@ -131,6 +142,24 @@ private void addHostMetrics(final List<Item> metricsList, final long dcId, final
             int isDedicated = (dr != null) ? 1 : 0;
             metricsList.add(new ItemHostIsDedicated(zoneName, zoneUuid, host.getName(), host.getUuid(), host.getPrivateIpAddress(), isDedicated));
 
+            List<String> hostTags = _hostTagsDao.gethostTags(host.getId());
+            String hosttags = StringUtils.join(hostTags, ",");
+            for (String tag : hostTags) {
+                Integer current = totalHosts.get(tag) != null ? totalHosts.get(tag) : 0;
+                totalHosts.put(tag, current + 1);
+            }
+            if (host.getStatus() == Status.Up) {
+                for (String tag : hostTags) {
+                    Integer current = upHosts.get(tag) != null ? upHosts.get(tag) : 0;
+                    upHosts.put(tag, current + 1);
+                }
+            } else if (host.getStatus() == Status.Disconnected || host.getStatus() == Status.Down) {
+                for (String tag : hostTags) {
+                    Integer current = downHosts.get(tag) != null ? downHosts.get(tag) : 0;
+                    downHosts.put(tag, current + 1);
+                }
+            }
```

Comment:
i feel i have seen this code before and made some suggestions to extract it.
```suggestion
            String hosttags = markTagMaps(totalHosts, upHosts, downHosts);
```
```
    private String markTagMaps(Map<String, Integer> totalHost, Map<String, Integer> upHost, Map<String, Integer> downHost) {
            List<String> hostTags = _hostTagsDao.gethostTags(host.getId());
            String hosttags = StringUtils.join(hostTags, ",");
            markTags(hostTags, totalHosts);
            if (host.getStatus() == Status.Up) {
                markTags(hostTags, upHosts);
            } else if (host.getStatus() == Status.Disconnected || host.getStatus() == Status.Down) {
                markTags(hostTags, downHosts);
            }
    }
```
and
```
    private void markTags(List<String> tags, Map<String, Integer> tagMap) {
                for (String tag : tags) {
                    Integer current = tagMap.get(tag) != null ? tagMap.get(tag) : 0;
                    tagMap.put(tag, current + 1);
                }
    }
```

further on n the method some extractions should be made as well

Reply:
Yeap, You have seen it in the https://github.com/apache/cloudstack/pull/5773 pr :)